src/main/java/org/testng/TestRunner.java

```diff
@@ -961,7 +970,18 @@ public ISuite getSuite() {
 
   @Override
   public ITestNGMethod[] getAllTestMethods() {
-    return m_allTestMethods;
+    if (m_allJunitTestMethods != null) {
+      //This is true only when we are running JUnit mode
+      return m_allJunitTestMethods;
+    }
+    if (m_groupMethods != null) {
+      //Be it memory friendly mode or normal mode, we still have a
+      // configuration object that remembers all the test methods. So we always query it
+      return m_groupMethods.getAllTestMethods();
```

Comment:
It is working but the design smells a bit here because there are 2 different purposes: 

1. The TestRunner "cache" should work with or without `m_groupMethods` (imagine we want to drop `getAllTestMethods` from `m_groupMethods`) ;
2. If the "cache" notion is shared between different objects, we can share it by a holder object and manage it in a single point. (it will be easier to follow the cache management)

Reply:
@juherr - So you are basically suggesting that we refactor both `TestRunner` and `ConfigurationGroupMethods` so that it uses a test method cache ? I can give that a try.

Reply:
@juherr - Just for the sake of argument, why cant we treat `ConfigurationGroupMethods ` as that cache ? This class will not go away for sure unless and until we have drastically changed TestNG behavior, and it needs all the test methods references. So why not just have it also play an additional role of being the cache without it being the cache ? Something like a side effect of its original working ? I am not quite sure as to why that would not make sense ? What additional benefit is the holder object going to give us (Apart from it being just a POJO) ?

Reply:
Your solution is working and it is just a naming issue: I'm not sure to understand what is the purpose of `ConfigurationGroupMethods`.
I mean `ConfigurationGroup` methods don't match well with `all` methods.

Then I've just checked and according to the documentation:
> This class wraps access to beforeGroups and afterGroups methods, since they are passed around the
> various invokers and potentially modified in different threads.

Clearly, its responsibility cannot be a cache. 
But a cache can be easily shared between classes without any problem.

Reply:
Fixed